File: android_app/lib/models/channel_data.dart
// ─────────────────────────────────────────────
// PocketTX – ChannelData Model
// Symmetric with Windows Companion ChannelData.
// Holds normalized + PWM values for 8 channels.
// ─────────────────────────────────────────────

import 'package:equatable/equatable.dart';
import '../core/constants/channel_constants.dart';

/// Immutable snapshot of all 8 RC channel values.
/// Schema version: 1 (see ModelVersions.channelData)
class ChannelData extends Equatable {
  final List<double> normalized; // 8 values, -1.0 to +1.0
  final List<int> pwm;           // 8 values, 1000us to 2000us
  final DateTime timestamp;
  final int schemaVersion;

  const ChannelData._({
    required this.normalized,
    required this.pwm,
    required this.timestamp,
    this.schemaVersion = 1,
  });

  /// Creates a ChannelData with all channels at center/idle.
  factory ChannelData.idle() {
    final List<double> norm = List.filled(ChannelConstants.channelCount, 0.0);
    // Throttle starts at min, not center
    norm[ChannelConstants.chThrottle] = -1.0;
    return ChannelData._(
      normalized: List.unmodifiable(norm),
      pwm: List.unmodifiable(norm.map(ChannelConstants.normalizedToPwm).toList()),
      timestamp: DateTime.now(),
    );
  }

  /// Creates ChannelData from normalized values list.
  factory ChannelData.fromNormalized(List<double> normalized) {
    assert(normalized.length == ChannelConstants.channelCount);
    final clamped = normalized
        .map((v) => v.clamp(ChannelConstants.normalizedMin, ChannelConstants.normalizedMax))
        .toList();
    return ChannelData._(
      normalized: List.unmodifiable(clamped),
      pwm: List.unmodifiable(clamped.map(ChannelConstants.normalizedToPwm).toList()),
      timestamp: DateTime.now(),
    );
  }

  /// Creates ChannelData from PWM values list.
  factory ChannelData.fromPwm(List<int> pwm) {
    assert(pwm.length == ChannelConstants.channelCount);
    final clampedPwm = pwm.map((p) => p.clamp(1000, 2000)).toList();
    final norm = clampedPwm.map(ChannelConstants.pwmToNormalized).toList();
    return ChannelData._(
      normalized: List.unmodifiable(norm),
      pwm: List.unmodifiable(clampedPwm),
      timestamp: DateTime.now(),
    );
  }

  /// Returns a copy with updated [index] channel value.
  ChannelData withChannel(int index, double normalizedValue) {
    assert(index >= 0 && index < ChannelConstants.channelCount);
    final updated = List<double>.from(normalized);
    updated[index] = normalizedValue.clamp(-1.0, 1.0);
    return ChannelData.fromNormalized(updated);
  }

  double get roll => normalized[ChannelConstants.chRoll];
  double get pitch => normalized[ChannelConstants.chPitch];
  double get throttle => normalized[ChannelConstants.chThrottle];
  double get yaw => normalized[ChannelConstants.chYaw];
  double get aux1 => normalized[ChannelConstants.chAux1];
  double get aux2 => normalized[ChannelConstants.chAux2];
  double get aux3 => normalized[ChannelConstants.chAux3];
  double get aux4 => normalized[ChannelConstants.chAux4];

  int get rollPwm => pwm[ChannelConstants.chRoll];
  int get pitchPwm => pwm[ChannelConstants.chPitch];
  int get throttlePwm => pwm[ChannelConstants.chThrottle];
  int get yawPwm => pwm[ChannelConstants.chYaw];

  Map<String, dynamic> toJson() => {
        'normalized': normalized,
        'pwm': pwm,
        'timestamp': timestamp.toIso8601String(),
        'schemaVersion': schemaVersion,
      };

  @override
  List<Object?> get props => [normalized, pwm, timestamp];

  @override
  String toString() =>
      'ChannelData(R:${roll.toStringAsFixed(2)} '
      'P:${pitch.toStringAsFixed(2)} '
      'T:${throttle.toStringAsFixed(2)} '
      'Y:${yaw.toStringAsFixed(2)})';
}

File: companion/src/PocketTX.Companion.Core/Models/ChannelData.cs
using PocketTX.Companion.Shared.Constants;

namespace PocketTX.Companion.Core.Models;

/// <summary>
/// Domain model encapsulating normalized (-1.0 to 1.0) and PWM (1000us to 2000us) channel values.
/// </summary>
public sealed class ChannelData
{
    public float[] NormalizedValues { get; }
    public ushort[] PwmValues { get; }
    public bool[] DigitalSwitches { get; }
    public DateTime Timestamp { get; set; }

    public ChannelData(int channelCount = ChannelConstants.DefaultChannelCount)
    {
        NormalizedValues = new float[channelCount];
        PwmValues = new ushort[channelCount];
        DigitalSwitches = new bool[channelCount];
        Timestamp = DateTime.UtcNow;

        // Default neutral positions
        NormalizedValues[0] = 0.0f; // Roll
        NormalizedValues[1] = 0.0f; // Pitch
        NormalizedValues[2] = -1.0f; // Throttle down
        NormalizedValues[3] = 0.0f; // Yaw

        UpdatePwmFromNormalized();
    }

    public void SetChannelNormalized(int index, float normalizedValue)
    {
        if (index < 0 || index >= NormalizedValues.Length) return;

        NormalizedValues[index] = Math.Clamp(normalizedValue, ChannelConstants.MinNormalized, ChannelConstants.MaxNormalized);
        PwmValues[index] = NormalizedToPwm(NormalizedValues[index]);
        if (index < DigitalSwitches.Length)
        {
            DigitalSwitches[index] = NormalizedValues[index] > 0.0f;
        }
        Timestamp = DateTime.UtcNow;
    }

    public void SetChannelPwm(int index, ushort pwmValue)
    {
        if (index < 0 || index >= PwmValues.Length) return;

        PwmValues[index] = Math.Clamp(pwmValue, ChannelConstants.MinPwmUs, ChannelConstants.MaxPwmUs);
        NormalizedValues[index] = PwmToNormalized(PwmValues[index]);
        if (index < DigitalSwitches.Length)
        {
            DigitalSwitches[index] = NormalizedValues[index] > 0.0f;
        }
        Timestamp = DateTime.UtcNow;
    }

    public void UpdatePwmFromNormalized()
    {
        for (int i = 0; i < NormalizedValues.Length; i++)
        {
            PwmValues[i] = NormalizedToPwm(NormalizedValues[i]);
        }
    }

    public static ushort NormalizedToPwm(float normalized)
    {
        float clamped = Math.Clamp(normalized, -1.0f, 1.0f);
        return (ushort)Math.Round(ChannelConstants.CenterPwmUs + (clamped * 500.0f));
    }

    public static float PwmToNormalized(ushort pwm)
    {
        ushort clamped = Math.Clamp(pwm, ChannelConstants.MinPwmUs, ChannelConstants.MaxPwmUs);
        return (clamped - (float)ChannelConstants.CenterPwmUs) / 500.0f;
    }
}
